### PR TITLE
6D Router Lookahead

### PR DESCRIPTION
**Pull request by amin1377**
In this pull request, the main data structure of the router lookahead is updated to have six dimensions instead of five. The previous dimensions were: [dest. layer][chanx/y][seg_type][width][height].

The new dimensions are: [source layer][chanx/y][seg_type][dest. layer][width][height].

Previously, since the connections were only from OPINs, each layer had a separate router lookahead. In the case of die crossing, the minimum delay of the inter-layer connection was stored and added to the cost to reach the sink on the destination layer from that point.

As development progresses in adding inter-die connections, we needed to update the router lookahead to consider that connections can cross die on switch blocks. To achieve this, an additional dimension is added to the array. Therefore, the first index related to the layer is now the source layer, and the second index related to the layer pertains to the layer on which the sink is located.

### CODE REVIEW

**Comment by amin1377** [comment]
Titan QoR:
Screenshot from 2023-11-23 08-42-25

**Comment by vaughnbetz** [comment]
Slight 1.6% router time increase, but since place time is up .5% and pack time 1.9% this is hopefully machine load noise.  In any case we need it so merging, and Amin is working on optimization of the overall lookahead size and compute time for the future (beyond this PR).

File: vpr/src/route/router_lookahead_map.h
Comment (vaughnbetz) [inline vpr/src/route/router_lookahead_map.h:61]: "layer_num" mentioned above --> target "layer_num" mentioned above

File: vpr/src/route/router_lookahead_map_utils.cpp
Comment (vaughnbetz) [inline vpr/src/route/router_lookahead_map_utils.cpp:27]: Say what the @return is

File: vpr/src/route/router_lookahead_map_utils.h
Comment (vaughnbetz) [inline vpr/src/route/router_lookahead_map_utils.h:297]: Should probably just get this from the router context, but your call.

**Review by vaughnbetz — APPROVED** [review]
Looks fine; a few minor commenting changes etc. to consider. Could be done in a separate PR if you wish or we can merge without waiting for the commenting CI to finish/pass.

File: vpr/src/route/router_lookahead_map_utils.h
Comment (amin1377) [reply vpr/src/route/router_lookahead_map_utils.h:297]: We don't have it in the router context yet.

**Comment by amin1377** [comment]
@vaughnbetz This PR is ready to merge

**Pull request merged**
